xds/src/main/java/io/grpc/xds/XdsRoutingLoadBalancer.java

```diff
@@ -16,18 +16,209 @@
 
 package io.grpc.xds;
 
+import static io.grpc.ConnectivityState.CONNECTING;
+import static io.grpc.ConnectivityState.IDLE;
+import static io.grpc.ConnectivityState.READY;
+import static io.grpc.ConnectivityState.TRANSIENT_FAILURE;
+import static io.grpc.xds.XdsSubchannelPickers.BUFFER_PICKER;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import io.grpc.ConnectivityState;
+import io.grpc.InternalLogId;
 import io.grpc.LoadBalancer;
+import io.grpc.MethodDescriptor;
 import io.grpc.Status;
+import io.grpc.internal.ServiceConfigUtil.PolicySelection;
+import io.grpc.util.ForwardingLoadBalancerHelper;
+import io.grpc.util.GracefulSwitchLoadBalancer;
+import io.grpc.xds.XdsLogger.XdsLogLevel;
+import io.grpc.xds.XdsRoutingLoadBalancerProvider.MethodName;
+import io.grpc.xds.XdsRoutingLoadBalancerProvider.Route;
+import io.grpc.xds.XdsRoutingLoadBalancerProvider.XdsRoutingConfig;
+import io.grpc.xds.XdsSubchannelPickers.ErrorPicker;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
 
-// TODO(zdapeng): Implementation.
 /** Load balancer for xds_routing policy. */
 final class XdsRoutingLoadBalancer extends LoadBalancer {
 
+  private final XdsLogger logger;
+  private final Helper helper;
+  private final Map<String, GracefulSwitchLoadBalancer> routeBalancers = new HashMap<>();
+  private final Map<String, RouteHelper> routeHelpers = new HashMap<>();
+
+  private Map<String, PolicySelection> actions = ImmutableMap.of();
+  private List<Route> routes = ImmutableList.of();
+
+  XdsRoutingLoadBalancer(Helper helper) {
+    this.helper = helper;
+    logger = XdsLogger.withLogId(
+        InternalLogId.allocate("xds-routing-lb", helper.getAuthority()));
+    logger.log(XdsLogLevel.INFO, "Created");
+  }
+
+  @Override
+  public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+    logger.log(XdsLogLevel.DEBUG, "Received resolution result: {0}", resolvedAddresses);
+    XdsRoutingConfig xdsRoutingConfig =
+        (XdsRoutingConfig) resolvedAddresses.getLoadBalancingPolicyConfig();
+    if (xdsRoutingConfig == null) {
+      helper.updateBalancingState(
```

Comment:
Should this also shut down child balancers (if exist) as well? 

Note you shouldn't call `handleNameResolutionError()` API directly. You could have a helper method for error handling and have `handleNameResolutionError()` call it as well.

Reply:
Why should this shutdown child balancers? Even `handleNameResolutionError()` does not shutdown child balancers.

Reply:
I mean propagate error to child policies. `handleNameResolutionError()` propagates error down to child policies and eventually have leaf policies propagate `TRANSIENT_FAILURE` back to root policy. But here, it behaves differently.

Reply:
Maybe we can just `checkNotNull` for  `xdsRoutingConfig` like `CdsLoadBalancer` does, it can be only programing error of our own.  I don't remember why `EdsLoadBalancer` needs `updateBalancingState` instead of `checkNotNull`.

Reply:
I am fine with either. But if you are updating balancing state, you'd better do it completely.